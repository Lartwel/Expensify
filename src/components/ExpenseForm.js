import React from 'react'
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize';

export default class ExpenseForm extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount/100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      id: props.expense ? props.expense.id : ''
    }
  }
  onDescriptionChange = (e)=>{
    const description = e.target.value
    this.setState({description})
  }
  onAmountChange = (e)=>{
    const amount = e.target.value;
    if ( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/) ){
      this.setState({amount})
    }
  }
  onNoteChange = (e)=>{
    const note = e.target.value
    this.setState({note})
  }
  onFocusChange = ({focused}) => {
    this.setState({calendarFocused: focused});
  }
  onDateChange = date => {
    return date && this.setState({createdAt: date})
  }

  onSubmit = e => {
    e.preventDefault();
    const {amount, description} = this.state;

    if( !amount || !description ){
      this.setState({error: 'Please provide description and amount.'})
    } else{
      this.setState({error: ''})
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(amount, 10) * 100, //cents
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf(), //melliseconds,
        id: this.state.id
      })
    }
  }

  render(){
    return (
      <form onSubmit={this.onSubmit}>
        <input 
          type="text"
          value={this.state.description}
          placeholder='Description'
          onChange={this.onDescriptionChange}
        />
        <input 
          type="text"
          value={this.state.amount}
          placeholder='Amount'
          onChange={this.onAmountChange}
        />

        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange} 
          focused={this.state.calendarFocused} 
          onFocusChange={this.onFocusChange}
          id="expense_date"
          numberOfMonths={1}
          isOutsideRange={() => false}
        />

        <textarea
          placeholder='Add a note for your expense (optional)'
          value={this.state.note}
          onChange={this.onNoteChange}
        ></textarea>
    

        <input type="submit" />

        {this.state.error && <div>
          <i>Error</i> {this.state.error}
        </div>}

      </form>
    )
  }
}

