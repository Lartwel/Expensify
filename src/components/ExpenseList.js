import React from 'react';
import {connect} from 'react-redux';
import ExpenseListeItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


export const ExpenseList = props => (
     props.expenses.length === 0 ? 
     (
        <div>No Expense Found</div>
     ) : 
     (
      <div>
        <h1>Expense List</h1>
        {
          props.expenses.map(expense => <ExpenseListeItem key={expense.id} {...expense} />)
        }
      </div>
    )
  
);

const mapStateToProps = state => {
  console.log(state.expenses);
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpenseList);

