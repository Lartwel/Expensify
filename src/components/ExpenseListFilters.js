import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilters extends React.Component{
  constructor(props){
    super(props);
    this.state = {
      calendarFocused: null
    }
  }
  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange = (calendarFocused => {
    this.setState({calendarFocused})
  })
  onTextChange =(e) => this.props.setTextFilter(e.target.value)
  onSortChange = e => {
    e.target.value === 'amount' ? this.props.sortByAmount() : this.props.sortByDate()
  }

  render(){
    return(
      <div>
        <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>

        <select 
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
          startDateId='start_date'
          endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
          endDateId='end_date'
          onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
          onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
          focusedInput={this.state.calendarFocused}
          isOutsideRange={() => false}
          numberOfMonths={1}
          showClearDates={true}
        />
      </div>
    )
  }
}

const mapStateToProps = state => ({
  filters: state.filters
})

const mapDispatchToProps = dispatch => ({
  setTextFilter: value => dispatch(setTextFilter(value)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: EndDate => dispatch(setEndDate(EndDate)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);