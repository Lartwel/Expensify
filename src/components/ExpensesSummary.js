import React from 'react';
import selectExpensesTotal from '../selectors/expenses-total';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import expensesTotalSelector from '../selectors/expenses-total';
import numeral from 'numeral';
import { connect } from 'react-redux';

export const ExpensesSummary = ({expensesTotal, expensesCount}) => {
  const expenseWord = expensesCount.length === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,00.00');
  return(
    <div className="page-header">
      <div className="content-container">
      <div className="page-header__title">
        Viewing <span>{expensesCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span>
      </div>
      <div className="page-header_actions">
        <Link to="/create-expense" className="button">Add Expense</Link>
      </div>
      </div>
    </div>
  )
}

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters)
  return{
    expensesTotal: expensesTotalSelector(visibleExpenses),
    expensesCount: visibleExpenses.length,
  }
}


export default connect(mapStateToProps)(ExpensesSummary)

