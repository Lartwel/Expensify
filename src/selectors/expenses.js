import moment from 'moment';

// Selectors

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter(expense => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return  textMatch && startDateMatch && endDateMatch;
  }).sort((a, b) => {
    return sortBy === 'date' ? a.createdAt - b.createdAt : (sortBy === 'amount' ? a.amount - b.amount : 0)
  })
};