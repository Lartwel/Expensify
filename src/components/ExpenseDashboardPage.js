import React from 'react';
import ConnectedExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => {
  return (
    <div>
      <ExpensesSummary/>
      <ExpenseListFilters/>
      <ConnectedExpenseList/>
    </div>
  )
}

export default ExpenseDashboardPage;
