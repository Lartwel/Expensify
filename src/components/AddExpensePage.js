import React, { Component } from 'react'
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses'

export class AddExpensePage extends Component {
  addExpense = expense => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <h1>Create Expense</h1>
        <ExpenseForm
          onSubmit={this.addExpense}
        />      
      </div>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)