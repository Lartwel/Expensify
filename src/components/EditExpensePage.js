import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component{
  constructor(props){
    super(props)
  }
  editExpense = expense => {
    this.props.editExpense(expense.id, expense)
    this.props.history.push('/');
  }
  removeExpense = () => {
    this.props.removeExpense({ id: this.props.expense.id })
    this.props.history.push('/');
  }
  render(){
    return(
      <div>
        <ExpenseForm
          onSubmit={this.editExpense}
          expense={this.props.expense}
          id="edit_expense_form"
        />

        <button onClick={this.removeExpense}>Remove</button>

      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find( expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: ( id, expense ) => dispatch( editExpense(id, expense) ),
  removeExpense: data => dispatch( removeExpense(data) )
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
