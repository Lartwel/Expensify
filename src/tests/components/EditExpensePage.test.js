import React from 'react';
import { EditExpensePage } from '../../components/EditExpensePage';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

let wrapper, editExpense, removeExpense, history;
beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<EditExpensePage 
    editExpense={editExpense} 
    removeExpense={removeExpense} 
    history={history} 
    expense={expenses[1]}
    /> );
})

test('should render Edit Expense Page correctly', () => {
  expect(wrapper).toMatchSnapshot();
})

test('should handle Edit Expense correctly', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])
})

test('should handle Remove Expense', () => {
  wrapper.find('button').simulate('click');
  expect(removeExpense).toHaveBeenLastCalledWith({
    id: expenses[1].id
  });
  // expect(history.push).toHaveBeenLastCalledWith('/');
})