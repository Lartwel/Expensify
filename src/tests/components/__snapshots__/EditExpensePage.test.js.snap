// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Edit Expense Page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <div
        className="page-header__title"
      >
        Edit Expense
      </div>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 100,
          "createdAt": -345600000,
          "description": "Electricity Bill",
          "id": "2",
          "note": "",
        }
      }
      id="edit_expense_form"
      onSubmit={[Function]}
    />
    <button
      className="button button-danger"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
